lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -2535,6 +2544,281 @@ void StaticRepeatOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
   patterns.add(zeroRepeat);
 }
 
+//===----------------------------------------------------------------------===//
+// InvokeOp
+//===----------------------------------------------------------------------===//
+
+// Parse the parameter list of invoke.
+static ParseResult
+parseParameterList(OpAsmParser &parser, OperationState &result,
+                   SmallVectorImpl<OpAsmParser::UnresolvedOperand> &ports,
+                   SmallVectorImpl<OpAsmParser::UnresolvedOperand> &inputs,
+                   SmallVectorImpl<Attribute> &portNames,
+                   SmallVectorImpl<Attribute> &inputNames,
+                   SmallVectorImpl<Type> &types) {
+  OpAsmParser::UnresolvedOperand port;
+  OpAsmParser::UnresolvedOperand input;
+  Type type;
+  auto parseParameter = [&]() -> ParseResult {
+    if (parser.parseOperand(port) || parser.parseEqual() ||
+        parser.parseOperand(input))
+      return failure();
+    ports.push_back(port);
+    portNames.push_back(StringAttr::get(parser.getContext(), port.name));
+    inputs.push_back(input);
+    inputNames.push_back(StringAttr::get(parser.getContext(), input.name));
+    return success();
+  };
+  if (parser.parseCommaSeparatedList(OpAsmParser::Delimiter::Paren,
+                                     parseParameter))
+    return failure();
+  if (parser.parseArrow())
+    return failure();
+  auto parseType = [&]() -> ParseResult {
+    if (parser.parseType(type))
+      return failure();
+    types.push_back(type);
+    return success();
+  };
+  return parser.parseCommaSeparatedList(OpAsmParser::Delimiter::Paren,
+                                        parseType);
+}
+
+ParseResult InvokeOp::parse(OpAsmParser &parser, OperationState &result) {
+  StringAttr componentName;
+  SmallVector<OpAsmParser::UnresolvedOperand, 4> ports;
+  SmallVector<OpAsmParser::UnresolvedOperand, 4> inputs;
+  SmallVector<Attribute> portNames;
+  SmallVector<Attribute> inputNames;
+  SmallVector<Type, 4> types;
+  if (parser.parseSymbolName(componentName))
+    return failure();
+  FlatSymbolRefAttr callee = FlatSymbolRefAttr::get(componentName);
+  SMLoc loc = parser.getCurrentLocation();
+  result.addAttribute("callee", callee);
+  if (parseParameterList(parser, result, ports, inputs, portNames, inputNames,
+                         types))
+    return failure();
+  if (parser.resolveOperands(ports, types, loc, result.operands))
+    return failure();
+  if (parser.resolveOperands(inputs, types, loc, result.operands))
+    return failure();
+  result.addAttribute("portNames",
+                      ArrayAttr::get(parser.getContext(), portNames));
+  result.addAttribute("inputNames",
+                      ArrayAttr::get(parser.getContext(), inputNames));
+  return success();
+}
+
+void InvokeOp::print(OpAsmPrinter &p) {
+  // print the parameter list.
```

Comment:
If you really want a comment here, it should be above the function definition, and not directly inside it.